test/functional/master/test_build_cancellation.py

```diff
@@ -0,0 +1,66 @@
+
+
+@skipIf(is_windows(), 'Fails on AppVeyor; see issue #345')
+class TestBuildCancellation(BaseFunctionalTestCase):
+    def test_build_cancellation_while_building(self):
+        master = self.cluster.start_master()
+        # Only one slave, with one executor. This means that the slave should be able to
+        # theoretically finish the build in 30 seconds, as this job definition has 30 atoms,
+        # with each sleeping for 1 second.
+        slaves = self.cluster.start_slaves(1, num_executors_per_slave=1, start_port=43001)
+        project_dir = tempfile.TemporaryDirectory()
+        build_resp = master.post_new_build({
+            'type': 'directory',
+            'config': yaml.safe_load(JOB_WITH_SLEEPS_90SEC.config[os.name])['SleepingJob90Sec'],
+            'project_directory': project_dir.name,
+        })
+        build_id = build_resp['build_id']
+        self.assertTrue(master.block_until_build_started(build_id, timeout=30),
+                        'The build should start building within the timeout.')
+        self.assert_build_status_contains_expected_data(build_id, {'status': 'BUILDING'})
+        master.cancel_build(build_id)
+        self.assertTrue(master.block_until_build_canceled(build_id, timeout=30),
+                        'The build should be canceled within the timeout.')
+        self.assert_build_status_contains_expected_data(build_id, {'status': 'CANCELED'})
+        # Make sure that slave becomes idle before the total job execution time (90 sec).
+        # Which means the build is canceled and aborted before it can finish.
+        # So here total maximum timeout we have is
+        #   30 (block_until_build_canceled)
+        # + 10 (block_until_idle)
+        # = 40 sec; which is less than the build execution time (i.e. 90 sec)
+        self.assertTrue(slaves[0].block_until_idle(timeout=10),
+                        'The slave should become idle within the timeout.')
+
+    def test_build_cancellation_while_atomizing(self):
+        master = self.cluster.start_master()
+        # Start a slave even though its not needed as build is canceled during atomization
+        slaves = self.cluster.start_slaves(1, num_executors_per_slave=1, start_port=43001)
+        project_dir = tempfile.TemporaryDirectory()
+        build_resp = master.post_new_build({
+            'type': 'directory',
+            'config': yaml.safe_load(JOB_WITH_SLEEPING_ATOMIZER_90SEC.config[os.name])['SleepingAtomizerJob90Sec'],
+            'project_directory': project_dir.name,
+        })
+        build_id = build_resp['build_id']
+        self.assertTrue(master.block_until_build_has_status(build_id, ['QUEUED'], timeout=30),
+                        'The build should be in queued state within the timeout.')
+        self.assert_build_status_contains_expected_data(build_id, {'status': 'QUEUED'})
+        # Sleep for 5 sec to let CR-master start the atomization process
+        time.sleep(5)
+        master.cancel_build(build_id)
```

Comment:
I was thinking that an alternate way to do this test (which would not require the `sleep(5)`) would be to make the atomizer something like:

`- ATOM: touch /tmp/atomizer-is-executing; for x in $(seq 1 90); do sleep 1 && echo $x; done`
Then the test could block until that /tmp/atomizer-is-executing file exists.

You could also do something like `echo "$$" > /tmp/atomizer-pid`, then have the test validate the atomizer process is killed, but I'm not sure if there's a good way to do that on windows.

My main concern with sleeps is that they can cause flaky test failures if the test runs slowly for some reason, and they slow down the test if the test would otherwise run quickly.